src/cascadia/TerminalControl/ControlCore.cpp

```diff
@@ -2119,6 +2286,10 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         auto context = winrt::make_self<CommandHistoryContext>(std::move(commands));
         context->CurrentCommandline(winrt::hstring{ _terminal->CurrentCommand() });
 
+        // TODO CARLOS: should we delete this after a new command is run? Or delete it after a suggestion is used? Or just after the next winget suggestion (current impl)?
+        //              No clue which we should do. Thoughts?
+        context->WinGetSuggestions(_cachedWinGetSuggestions);

```

Comment:
@zadjii-msft thoughts on how to proceed here? I think...
- **delete after a new command is run**: doable, but I'm not familiar with shell integration enough. If we wanna go this route, tips?
- **delete after a suggestion is used**: this doesn't sound like something we can do easily. And also it doesn't sound like it'd feel totally right, imo.
- **delete after the next winget suggestion**: I like that we don't hold on to stale suggestions, but I don't like that it's held on after another command is run.

Reply:
Okay this is gonna sound like maybe a wacky idea, but what if we just stored these on the buffer, on the ROW itself? Kinda a bit like what I'm doing over in https://github.com/microsoft/terminal/pull/16937/files#diff-af0f626e5bb6718b60aa712a64c969e274f30968b5e24cc013318cf5c6badc12. If we just have one collection of suggestions per-row (a `std::vector<wstring>`), then we'd be able to just iterate up to the most recent collection of quick fixes, and only return those. 

BUT even as I'm typing that up - I'd guess that a vector of strings, even an empty one, isn't low enough overhead to have the buffer go fast. 

I guess a map of row->vector<hstring> would work, but you'd have to deal with reflowing it and decrementing all rows on circling. Probably not terrible. 

Knowing the row that the suggestion was printed on would also make drawing the icons more trivial. 

another alternative / future thought - instead of just returning the most recent quick fix, you could further refine to be "the most recent quick fix if it was emitted in the most recent command _with output_". The most recent mark probably doesn't have output (it's just a prompt & possibly typed command), so keep looking backwards till you find one with output. Then only return quick fixes in _that_ mark's extent.

Reply:
Hmm... I'm leaning towards this idea:
> I guess a map of row->vector would work, but you'd have to deal with reflowing it and decrementing all rows on circling. Probably not terrible.

We need to do this now so that we don't have the same suggestions appear forever. But the annoying thing is going to be that until quick fixes is implemented, we're just gonna hold on to the winget suggestions until they the buffer circles with no way to access them (since sxn ui can only be applied to the input line, right?). We can probably add some code in the interim that forces the map to only hold on to the most recent one, then undo that behavior as a part of quick fixes.

Another thing to point out here is that when the VT sequence is emitted, it'll be at the beginning of the "is not recognized..." line (where the flag is drawn in the image below)
![selection marker at the beginning of CMD's command not found output](https://github.com/microsoft/terminal/assets/11050425/6aa6f450-bbd5-4e06-b473-ca9d0348044d)

So, this means that if there's multiple "is not recognized" messages, output from a command is going to have to consolidate all of them to suggest them as a quick fix on the next prompt. But without shell integration, we won't know which ones to consolidate.

As for this:
> another alternative / future thought - instead of just returning the most recent quick fix, you could further refine to be "the most recent quick fix if it was emitted in the most recent command with output". The most recent mark probably doesn't have output (it's just a prompt & possibly typed command), so keep looking backwards till you find one with output. Then only return quick fixes in that mark's extent.

I like that idea _a lot_, but we can't rely on scroll marks existing since it requires changes to the shell, right?

So, if we're going to need shell integration anyways, perhaps the "the most recent quick fix if it was emitted in the most recent command with output" idea is best then. It just feels weird to couple it with shell integration 🤔

Reply:
> we're just gonna hold on to the winget suggestions

oh no a couple strings, oh nooooo /s. Honestly, if it's the right design in the next ~2 releases timeframe, then let's just do it right and not try and over-engineer the interim solution. 

> So, this means that if there's multiple "is not recognized" messages, output from a command is going to have to consolidate all of them to suggest them as a quick fix on the next prompt. But without shell integration, we won't know which ones to consolidate

If we store them as a property on the row, then we can do:
* If there's a most recent prompt scroll mark, then get _all_ the quick fixes from all the ROWs until that prompt
* There's no shell integration? Then just get the quick fixes on the most recent ROW with quick fixes. 

Basically, it works without shell integration, but it could work _better_ with it